kernel/libs/aster-bigtcp/src/iface/ext.rs

```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MPL-2.0
+
+/// Extension to be implemented by users of this crate.
+pub trait Ext {
```

Comment:
Could you please try to give the trait a more specific name? For example, `NextPoll` or `ScheduleNextPoll`? If the trait may be extended in the future, then the trait name can be renamed accordingly.

**EDIT**: I see you add two associate observer type in the second commit. Maybe the name should be reconsidered, how about `SocketEventExt`?

Reply:
The problem is that this trait is referenced in too many places in the bigtcp crate. All interfaces and all sockets use it. So I think it deserves a short name, otherwise it's a pain to write a longer name as the trait bound to so many types.

```
> find . -name "*.rs" | xargs grep ': Ext' | wc -l
56
```

We need such a type so that the crate user (`aster-nix`) can provide some helper types (e.g. event observers)/methods (e.g. schedule the next poll) for bigtcp. This type has to be a "big" type, because it's always annoying to write multiple generic parameters (and have multiple trait bounds). So I'd like to avoid splitting the "big" type into several small types, even though each small type may have clear meanings.

Reply:
Thanks for your explanation! Yes, I agree it may be better to not have multiple traits in generic bounds. 

But when I see something like `T: Send+Sync+Ext`, I always feel confused that what `Ext` stands for. How do you like just using `iface::Ext` for everywhere you want to reference `Ext`? I saw you already introduce such pattern in this PR like `impl iface::Ext for IfaceExt `. It will looks much clearer for me.

Reply:
I think the confusion comes from mixing the extension for an iface (aka `ScheduleNextPoll` in your suggested naming) and the extension for the entire crate (I'd like to keep the `Ext` naming because it's short).

So I've tried to split them, and the result is the following API:

```rust
// In `aster-bigtcp/ext.rs`:

/// Extension to be implemented by users of this crate.
///
/// This should be implemented on an empty type that carries no data, since the type will never
/// actually be instantiated.
///
/// The purpose of having this trait is to allow users of this crate to inject multiple types
/// without the hassle of writing multiple trait bounds, which can be achieved by using the types
/// associated with this trait.
pub trait Ext {
    /// The type for ifaces to schedule the next poll.
    type ScheduleNextPoll: ScheduleNextPoll;

    /// The type for TCP sockets to observe events.
    type TcpEventObserver: SocketEventObserver;

    /// The type for UDP sockets to observe events.
    type UdpEventObserver: SocketEventObserver;
}
```

```rust
// In `aster-bigtcp/iface/sched.rs`:

/// A trait to provide the `schedule_next_poll` method for ifaces.
pub trait ScheduleNextPoll: Send + Sync {
    /// Schedules the next poll at the specific time.
    ///
    /// This is invoked with the time at which the next poll should be performed, or `None` if no
    /// next poll is required. It's up to the caller to determine the mechanism to ensure that the
    /// next poll happens at the right time (e.g. by setting a timer).
    fn schedule_next_poll(&self, ms: Option<u64>);
}
```

Could you please review the latest changes to see if they're better and/or address your comments?

Reply:
Yes, having only one `Ext` makes the code much clearer to me! The latest change looks fine to me!